### ISSUE TITLE
How does Oauth2 (Bearer token) work when getting access_token?

### ISSUE THREAD

**Issue by BarryThrill** [tiangolo/fastapi#801]
### First check

* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

I will start off by saying I have followed a guide of how to make a secure API using FastAPI (https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/).

What happens is that you will have two api which one is post and get

POST is where you pass the

`curl -X POST "http://127.0.0.1:8000/token" -H "accept: application/json" -H "Content-Type: application/x-www-form-urlencoded" -d "grant_type=&username=hello&password=world&scope=&client_id=&client_secret="`

GET is an api that requires the access_token Bearer to be able to access the API etc http://127.0.0.1:8000/getUser

Now I am pretty confused on how safe you can run the application because what makes me confused is that I would need to make a POST request with my username and password to the API which will return me access_token and I believe with that access_token I later on access the next api which is http://127.0.0.1:8000/getUser? Is this the correct way or am I out of scope?

Because what makes me unsafe is that if I now lets say "hardcode" my username and password into a python script and someone manages to reverese engineer the script/exe/whatever (Maybe even see through network what is sending as params/data)- they will be able to access my token in that case. So what would be the suggestion here?

**Comment by shifr** [comment]
The suggestion would be -  do not hardcode your username/password :)

**Comment by BarryAl-Jawari** [comment]
> The suggestion would be - do not hardcode your username/password :)

Indeed, I assume a better way is to connect to etc mongodb where you store your username and password instead for example?

**Comment by BarryThrill** [comment]
> The suggestion would be - do not hardcode your username/password :)

Indeed, I assume a better way is to connect to etc mongodb where you store your username and password instead for example?

**Comment by shifr** [comment]
> Indeed, I assume a better way is to connect to etc mongodb where you store your username and password instead for example?

Sure, in the example https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/ you can see `fake_users_db` - it's like a projection of the real database where you can store your users, but simplified for the tutorial purposes.

**Comment by BarryThrill** [comment]
> > Indeed, I assume a better way is to connect to etc mongodb where you store your username and password instead for example?
> 
> Sure, in the example https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/ you can see `fake_users_db` - it's like a projection of the real database where you can store your users, but simplified for the tutorial purposes.

Indeed will add that for sure! However I am not sure how it will handle when I try do a POST requests to /token. Will users see username and password through networks logs if I for etc. Upload it in heroku? Because that would suck haha. 

Because I need to send username and pass to token to get Token which I later on use that token to other api.

**Comment by shifr** [comment]
And also, as you can see in tutorial, you have to store `hashed passwords` instead of just raw passwords from input.

**Comment by BarryThrill** [comment]
> And also, as you can see in tutorial, you have to store `hashed passwords` instead of just raw passwords from input.

Correct. But when I do a POST requests to api/token then I need to send the data as etc {username: hello, password: world}

but that would be shown in networks logs if I would send a POST requests. Right!

**Comment by shifr** [comment]
> Because I need to send username and pass to token to get Token which I later on use that token to other api.

Yeah, that's the idea. You need one endpoint that accepts username and password as is, you must not to store this data or log somewhere.
Simple steps are:
- receive username/password
- generate the password hash
- get user from the database by username
- compare generated hash with a hash from the database
- if hashes are equal - return a token

**Comment by shifr** [comment]
> > And also, as you can see in tutorial, you have to store `hashed passwords` instead of just raw passwords from input.
> 
> Correct. But when I do a POST requests to api/token then I need to send the data as etc {username: hello, password: world}
> 
> but that would be shown in networks logs if I would send a POST requests. Right!

AFAIK, only client who do request can see it's payload and it will not be visible in logs, unless you put it there.

**Comment by BarryThrill** [comment]
> > > And also, as you can see in tutorial, you have to store `hashed passwords` instead of just raw passwords from input.
> > 
> > 
> > Correct. But when I do a POST requests to api/token then I need to send the data as etc {username: hello, password: world}
> > but that would be shown in networks logs if I would send a POST requests. Right!
> 
> AFAIK, only client who do request can see it's payload and it will not be visible in logs, unless you put it there.

Oh okey because I am not sure if I am thinking correctly now but my plan is to have two POST api

One for token which generates the token
Second for validate if a user is in my database or not

My plan was in my script which other people will have access (but will use pyarmor so they wont see the code) - In the start of the script I would send a post requests to the token api where I get the token and after that I send to my other api /validate that will see if a user is in my db or not (just bool of true or false)

What do you think about that? Am I thinking right or am I missing the point?

**Comment by shifr** [comment]
@BarryThrill hmm, why do you need to make second check if a user is in your db? When you're generating a token you're checking it, in one request. See tutorial `async def login_for_access_token` function

**Comment by BarryThrill** [comment]
> @BarryThrill hmm, why do you need to make second check if a user is in your db? When you're generating a token you're checking it, in one request. See tutorial `async def login_for_access_token` function

So I am not sure if my question fits in really in stackoverflow or not but I would give it a shot and try to see if my knowledge of JWT is actually correct or am I out of the loop totally.

So what I have created is an server API that reads POST requests that is being sent from a client app and that returns Bearer Token which is needed to be able to access rest of the API's that I have created.

So far I have a server api that created Bearer token IF the username and password matches the login.

A simple POST requests would look like

`    {'username': 'hello', 'password': 'world'}`

so what I have done instead is that I created a JWT encoded from JWT.IO site with a secret code which would look like:

```
    {'username': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImJhcnJ5In0.-TCwkrPr8dq4WqsckaWNG7G2ddn7e97hH0jkQ-1j5Bo',
     'password': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXNzd29yZCI6ImF1dG9zbmtyIn0.mWvxW4xga_OQMLKxf5zfSP4bSV0KzLPSRpqapU-RbAw'}

```
However my main problem is that how should I handle the connection between client app -> first /token request to be able to get bearer token?

It seems like I do need to somehow "hardcode" the username and password inside the client app to be able to access my API but I feel like this is not the correct way because then you would just be able to read the network logs and send the same requests to the server and you would forever get a new Bearer token which you can manipulate my server api.


My question is that what way should I do to not be able to expose my username and password in the client app AND to be able manipulate through my server api? Because what my server does is that it decodes the username and password from JWT with secret and matches if the username and password matches my server api username and password. But it feels like by exposing my username and password with already finished encoded JWT token you would still be able to use those values and do whatever you want?


Example of a client app:

```
    import requests
    
    headers = {
        'accept': 'application/json',
        'Content-Type': 'application/x-www-form-urlencoded',
    }
    
    data = {'username': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImJhcnJ5In0.-TCwkrPr8dq4WqsckaWNG7G2ddn7e97hH0jkQ-1j5Bo',
            'password': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXNzd29yZCI6ImF1dG9zbmtyIn0.mWvxW4xga_OQMLKxf5zfSP4bSV0KzLPSRpqapU-RbAw'}
    
    
    response = requests.post('http://127.0.0.1:8000/token', headers=headers, data=data, verify=False)


```

**Comment by kuwv** [comment]
@BarryThrill, I'd recommend not making assumptions about JWT or OAuth and stick only to proven methods you can reproduce.

IMO the password auth flow is fine when your identity provider and app are the same system. It is common for username and passwords to be base64 encoded in transit through TLS. Once you have retrieved your token you will present it for each subsequent request. The key thing here is that the browser client, api, and IdP should all be your system in this scenario and therefore trusted. So issuing the token doesn't work on other systems anyway, and passwords on your own system can be handled with effective audit practices.

If you want to publish the client on something like Google Play or IOS App Store then you should consider PKCE instead of hard-coded passwords. PKCE requires authorization_code instead of password flow though.

**Comment by tiangolo** [comment]
Thanks for the help here @shifr and @kuwv !

@BarryThrill About the flow, yes: a client sends `username` and `password` to a single path, and receives an access token. Then that client uses that access token to send requests to *multiple* paths (that provide the actual value of your API). That token expires (the password doesn't), so if the token is stolen, at least it won't work forever. Then when it is expired, the client sends the username and password again to the first path to get a new access token.

About someone reading the username or password in transit, you should have HTTPS, there's no much way around. Heroku probably has something to do that for you. If you deploy it yourself, you can get free HTTPS certificates with Let's Encrypt. For example using Traefik.

About where you store those credentials (passwords), as @shifr and @kuwv say, you would have your application in a provider you trust, in an environment (VPS server, etc) that you control, not in some server that others have access to. If you can't trust the server where your app will run, you probably can't deploy it there (nor anything else in that server).

If you can't use a database and your code is public (e.g. stored in a GitHub public repo) you could store credentials in environment variables and read them from those environment variables in your app. Most server providers will let you set environment variables even if you don't have access to a database.
